torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1298,3 +1352,25 @@ def _alloc_storage(data: torch.Tensor, size: torch.Size) -> None:
         data.storage().size() == 0
     ), "Then tensor storage should have been resized to be 0."
     data.storage().resize_(size.numel())  # type: ignore[attr-defined]
+
+
+def _calc_grad_norm(parameters: List[torch.nn.Parameter], p: float) -> torch.Tensor:
+    r"""Calculate gradient norm of an iterable of parameters.
+    Returns:
+        Total norm of the parameters (viewed as a single vector).
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    parameters = list(filter(lambda par: par.grad is not None, parameters))
+
+    if len(parameters) == 0:
+        return torch.tensor(0.0)
+    p = float(p)
+    if p == inf:
+        local_norm = max(par.grad.detach().abs().max() for par in parameters)  # type: ignore
+    else:
+        # Compute the norm in full precision no matter what
+        local_norm = torch.linalg.norm(
+            torch.stack([torch.linalg.norm(par.grad.detach(), p, dtype=torch.float32) for par in parameters]), p
+        ).to(dtype=parameters[0].dtype)  # type: ignore
+    return local_norm
```

Comment:
probably worth having a basic unittest for `_calc_grad_norm`. Is there also no other helper/utility that will return the gradient norm of a given set of parameters? There doesn't appear to be any FSDP specific logic here so we could separate it out into a more generally accessible folder in the future.

Reply:
Totally, which folder or directory do we feel is a better place?